File: eeschema/GOST-doc-gen/odfpy_iface.cpp
/*
 * This program source code file is part of KiCad, a free EDA CAD application.
 *
 * Copyright (C) 2013 Alexander Lunev <al.lunev@yahoo.com>
 * Copyright (C) 2013 KiCad Developers, see AUTHORS.txt for contributors.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, you may find one here:
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 * or you may search the http://www.gnu.org website for the version 2 license,
 * or you may write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 */

/**
 * @file odfpy_iface.cpp
 */

#include <macros.h>

#include <common_funcs.h>

#include <odfpy_iface.h>

namespace GOST_DOC_GEN {

ODFPY_IFACE::ODFPY_IFACE()
{
    // launch the Python interpreter
    Py_Initialize();
}


ODFPY_IFACE::~ODFPY_IFACE()
{
    Py_Finalize();
}


bool ODFPY_IFACE::Connect()
{
    wxString path = GetResourceFile( wxT( "odfpy_iface.py" ) );

    if( path == wxEmptyString )
        return false;

    // remove filename from the full path
#if defined (__WXMSW__)
    path.Replace( wxT( "\\odfpy_iface.py" ), wxEmptyString );
#else
    path.Replace( wxT( "/odfpy_iface.py" ), wxEmptyString );
#endif

    path = wxT( "import sys\nsys.path.append('" ) + path + wxT( "')" );

    if( PyRun_SimpleString( TO_UTF8( path ) ) == -1 )
    {
        wxMessageBox( PyErrStringWithTraceback(),
                      wxEmptyString,
                      wxOK | wxICON_ERROR );
        return false;
    }

    PyObject* pyModuleString = PyString_FromString( (char*)"odfpy_iface" );
    PyObject* pyModule = PyImport_Import( pyModuleString );

    if( !pyModule )
    {
        wxMessageBox( PyErrStringWithTraceback(),
                      wxEmptyString,
                      wxOK | wxICON_ERROR );
        return false;
    }

    m_pyOdfpy_iface_inst = PyObject_GetAttrString( pyModule, (char*)"odfpy_iface_inst" );

    return true;
}


void ODFPY_IFACE::Disconnect()
{
}


bool ODFPY_IFACE::LoadDocument( wxString aUrl )
{
    PyObject* pyResult = PyObject_CallMethodObjArgs( m_pyOdfpy_iface_inst,
                                                     PyString_FromString( (char*)"LoadDocument" ),
                                                     PyString_FromString( aUrl ),
                                                     NULL );

    return ( pyResult != NULL );
}


bool ODFPY_IFACE::AppendDocument( wxString aUrl )
{
    PyObject* pyResult = PyObject_CallMethodObjArgs( m_pyOdfpy_iface_inst,
                                                     PyString_FromString( (char*)"AppendDocument" ),
                                                     PyString_FromString( aUrl ),
                                                     NULL );

    return ( pyResult != NULL );
}


void ODFPY_IFACE::SelectTable( int aIndex )
{
    PyObject* pyResult = PyObject_CallMethodObjArgs( m_pyOdfpy_iface_inst,
                                                     PyString_FromString( (char*)"SelectTable" ),
                                                     PyInt_FromLong( aIndex ),
                                                     NULL );

    if( pyResult == NULL )
    {
        wxMessageBox( wxT( "ODFPY_IFACE: Unable to select table" ),
                      wxEmptyString,
                      wxOK | wxICON_ERROR );
    }
}


void ODFPY_IFACE::PutCell( wxString aCellAddr,
                           wxString aStr,
                           int      aStyle )
{
    PyObject* pyResult = PyObject_CallMethodObjArgs( m_pyOdfpy_iface_inst,
                                                     PyString_FromString( (char*)"PutCell" ),
                                                     PyString_FromString( aCellAddr ),
                                                     PyString_FromString( aStr ),
                                                     PyInt_FromLong( aStyle ),
                                                     NULL );

    if( pyResult == NULL )
    {
        wxMessageBox( wxT( "ODFPY_IFACE: Unable to put cell" ),
                      wxEmptyString,
                      wxOK | wxICON_ERROR );
    }
    Py_Finalize();

}

} // namespace GOST_DOC_GEN

File: eeschema/GOST-doc-gen/odfpy_iface.py
#!/usr/bin/env python

# This program source code file is part of KiCad, a free EDA CAD application.
#
# Copyright (C) 2013 Alexander Lunev <al.lunev@yahoo.com>
# Copyright (C) 2013 KiCad Developers, see AUTHORS.txt for contributors.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, you may find one here:
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
# or you may search the http://www.gnu.org website for the version 2 license,
# or you may write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA


#import os
import odf.opendocument
from odf.text import P
from odf.table import *
from odf.style import Style, ParagraphProperties, TextProperties

class ODFPY_IFACE():

    def __init__( self ):
        self.document = None
        self.curTable = None


    def LoadDocument( self, aUrl ):
        self.document = odf.opendocument.load( aUrl )
        #SelectTable( 0 )
        self.document.save('/home/a-lunev/git/bzr/GOST-doc-gen/eeschema/GOST-doc-gen/templates/other.odt')


    def AppendDocument( self, aUrl ):
        anotherDoc = odf.opendocument.load( aUrl )

        # Copy all properties from anotherDoc to the document
        for meta in anotherDoc.meta.childNodes[:]:
            self.document.meta.addElement( meta )

        for font in anotherDoc.fontfacedecls.childNodes[:]:
            self.document.fontfacedecls.addElement(font)

        for style in anotherDoc.styles.childNodes[:]:
            self.document.styles.addElement( style )

        for masterstyle in anotherDoc.masterstyles.childNodes[:]:
            self.document.masterstyles.addElement( masterstyle )

        for autostyle in anotherDoc.automaticstyles.childNodes[:]:
            self.document.automaticstyles.addElement( autostyle )

        for setting in anotherDoc.settings.childNodes[:]:
            self.document.settings.addElement( setting )

        self.document.text.addElement( anotherDoc.text.getElementsByType( Table )[0] )

        self.document.save('/home/a-lunev/git/bzr/GOST-doc-gen/eeschema/GOST-doc-gen/templates/other.odt')


    def SelectTable( self, aIndex ):
        self.curTable = self.document.text.getElementsByType( Table )[aIndex]


    def PutCell( self, aCellAddress, aStr, aStyle ):
        print aCellAddress, aStr
        #col_i = ord( aCellAddress[0] ) - ord( 'A' )
        #print col_i
        #row_i = int( aCellAddress[1:] )
        #print row_i
        col_i = 2
        row_i = 27
        row = self.curTable.getElementsByType( TableRow )[row_i]
        cell = row.getElementsByType( TableCell )[col_i]
        textFound = False
        for p in cell.getElementsByType( P ):
            for p_data in p.childNodes:
                if p_data.tagName == 'Text':
                    textFound = True
                    p_data.data = '123' #aStr #text.decode('string_escape')

        if( not textFound ):
            p.addText( '123' )

        self.document.save('/home/a-lunev/git/bzr/GOST-doc-gen/eeschema/GOST-doc-gen/templates/other.odt')


odfpy_iface_inst = ODFPY_IFACE()
